### PR TITLE
Block spacing: add axial gap block support

### LINKED ISSUE #34529: Design Tools: Enhance Block Gap control to support axial configuration
## What problem does this address?

#32366 adds a theme.json mechanism to provide a a `blockGap` value to space out items in a container. Spacing using gap has the benefit of letting you not worry about the first or last elements in a container, or complex nth-child margins in containers that wrap.

#33991 added a basic UI for it, with a single CSS variable to begin with.

## What is your proposed solution?

Like how the padding control offers axial controls, a similar interaction could be used to split the single `gap` value into a `column-gap` and `row-gap`:

padding

The control would still be subject to design improvements pending from #27331 work, but the primary interaction would be the same. Sucha  control would benefit Navigation, Social Icons and Buttons block, as those get gap support (#34525, #34526, #34527).

### PR DESCRIPTION
**Pull request by ramonjd**
## Description

Gap support for blocks was introduced in https://github.com/WordPress/gutenberg/pull/33991. 

It added a CSS var, `--wp--style--block-gap`, to store the value.

This PR splits gap support values into row and column values, and refactors the value of `--wp--style--block-gap` to employ shorthand syntax: `-wp--style--block-gap: 99px 1px;`. 

To house row and columns values, we're adding two CSS vars in addition to `--wp--style--block-gap`:

`--wp--style--block-row-gap`
`--wp--style--block-column-gap`

More CSS vars? Well, yes.

Splitting values means we can specify gap values for individual properties, offering us more control over vertical and horizontal spacing, and allows us to align other spacing definitions, such as margin, to the gap values –

`margin-left: var(--wp--style--block-column-gap, 2em);`

– see for example the [Columns Block styles.scss](https://github.com/WordPress/gutenberg/blob/97406be48180cbae3e51f2d63c96b5dd6cc41a9e/packages/block-library/src/columns/style.scss#L32). 

The persistence of `--wp--style--block-gap` seems tautological, yet, aside from retaining backwards compatibility, it grants us convenient access to the gap shorthand value.

Fixes https://github.com/WordPress/gutenberg/issues/34529
Stealing ideas from https://github.com/WordPress/gutenberg/pull/35301
Related https://github.com/WordPress/gutenberg/pull/34546

https://user-images.githubusercontent.com/6458278/140844813-87cc644f-3efe-49f3-a443-7159724e7877.mp4

## How has this been tested?

Test scenarios:

- [ ] Before opting-in to blockGap support on your theme, check that the defaults work. Insert a column block with 3 columns. In desktop, make sure the columns after the first child have `margin-left: var(--wp--style--block-column-gap, 2em);`
- [ ] Ensure that blocks that use one of the variables for margins, paddings or otherwise work as expected:
    - [ ] Columns
    - [ ] Navigation
    - [ ] Button
- [ ] Changes to `--wp--style--block-gap` should updates values in `--wp--style--block-[row|column]-gap` and vice versa.
- [ ] Update `navigation/block.json` `blockGap` support to `"blockGap": [ "vertical", "horizontal" ],` to test axial gap support.
    - [ ] When the row and gap values and units are the same, the `--wp--style--block-gap` CSS var should be a single value, e.g., `101px`
    - [ ] When the row and gap values are different, or they use different units, the `--wp--style--block-gap` CSS var should use shorthand syntax value, e.g., `101px 101%`
- [ ] A theme that uses `--wp--style--block-gap` to determine spacing should not be impacted when axial controls are turned on.
- [ ] Paste a block in the editor with existing `blockGap` values (See **Example block code** below )and make sure that block gap value is filled in the supports control in the sidebar.
<details>
  <summary>Example block code</summary>

```html
<!-- wp:columns {"style":{"spacing":{"blockGap":"169px"}}} -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>1</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>2</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>3</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```
</details>

Run the unit tests:

`npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/spacing-test.php`

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: lib/theme.json [hunk 1]
```diff
@@ -213,8 +213,8 @@
 			]
 		},
 		"spacing": {
-			"blockGap": null,
-			"margin": false,
+			"blockGap": true,
+			"margin": true,
```
Comment (ramonjd) [inline lib/theme.json]: Just for ease of testing. I'll revert this.

File: packages/block-library/src/button/style.scss [hunk 2]
```diff
@@ -50,15 +50,15 @@ $blocks-block__margin: 0.5em;
 	}
 
 	&.wp-block-button__width-25 {
-		width: calc(25% - (var(--wp--style--block-gap, #{$blocks-block__margin}) * 0.75));
+		width: calc(25% - (var(--wp--style--block-column-ap, #{$blocks-block__margin}) * 0.75));
```
Comment (andrewserong) [inline packages/block-library/src/button/style.scss]: Just noticed a tiny typo while skimming this PR 🙂
```suggestion
		width: calc(25% - (var(--wp--style--block-column-gap, #{$blocks-block__margin}) * 0.75));
```

File: packages/edit-site/src/components/global-styles/use-global-styles-output.js [hunk 3]
```diff
@@ -318,7 +318,7 @@ export const toStyles = ( tree, blockSelectors ) => {
 	const nodesWithSettings = getNodesWithSettings( tree, blockSelectors );
 
 	let ruleset =
-		'.wp-site-blocks > * { margin-top: 0; margin-bottom: 0; }.wp-site-blocks > * + * { margin-top: var( --wp--style--block-gap ); }';
+		'.wp-site-blocks > * { margin-top: 0; margin-bottom: 0; }.wp-site-blocks > * + * { margin-top: var( --wp--style--block-column-gap ); }';
```
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/use-global-styles-output.js]: Should this one be `row` gap since it's being used for `margin-top`?

**Review by andrewserong — COMMENTED** [review]
Thanks so much for picking up this exploration @ramonjd! Axial controls for blockGap are pretty challenging to reason about, but I like the way you've taken the grouped value and are using it to sync to the separate column / row values, it feels like a pragmatic step forward 👍

This is testing pretty well for me on the Buttons block, opting it in to the vertical and horizontal controls:



Something I noticed when running TwentyTwentyTwo theme is that it sets in its global styles the spacing of [`1.5rem` in its theme.json](https://github.com/WordPress/twentytwentytwo/blob/b585df3e8b6802fdf2deadda9bc44948b389a894/theme.json#L343) which gets rendered out on the server to `--wp--style--block-gap` as a single CSS variable. I think this means that we might need additional logic in `WP_Theme_JSON_Gutenberg` for rendering out the body spacing variables ([I think this is where it's set](https://github.com/wordpress/gutenberg/blob/0e7298080c4d6b3f80084ed4a641193544c65fbb/lib/class-wp-theme-json-gutenberg.php#L247)), to also generate the other two variables when a string value is set?

image

I imagine this would be similar logic to what you've included for the inline variable rendering for individual blocks?

Without it, the spacing between paragraphs, for example, doesn't get applied because it can't find the row variable:



I was also wondering if it'd be worth storing the value of the variable as a single value when the row/column values are the same (E.g. `1rem` instead of `1rem 1rem`)?

But overall, I'm pretty excited about the progress you're making here — it demonstrates that it'll be possible to use axial controls while also not being _totally_ locked to CSS `gap` as `margin` can still use the individual row-gap value on its own 👍

And it could be pretty cool to see what sorts of patterns the axial gap might enable in WP 6.0 — perhaps a tile-based Query Loop pattern that is tucked in closely horizontally, but has a larger vertical gap between blog post cards?

Happy to do more testing next week! 😃

**Comment by ramonjd** [comment]
I'm very grateful for review @andrewserong 🙇 

> Without it, the spacing between paragraphs, for example, doesn't get applied because it can't find the row variable:

Great find, and thanks for the advice. I'll take a look at this.

> I was also wondering if it'd be worth storing the value of the variable as a single value when the row/column values are the same (E.g. 1rem instead of 1rem 1rem)?

It would look nicer, and probably be less confusing. 😄 Is there any benefit otherwise?

**Comment by andrewserong** [comment]
> It would look nicer, and probably be less confusing. 😄 Is there any benefit otherwise?

It might just improve neatness. Originally, when I first looked at the PR I thought we might need the single value instead of storing both values in the variable for cases where the CSS variable is used on a margin property (e.g. `margin-top` or `margin-bottom` in the layout block support). However, since you've updated the block support to use the row gap individual value instead, that doesn't appear to be an issue. So, I think it's probably just for neatness 🙂

In terms of the API we're effectively creating via the CSS variables, I suppose for other blocks that might want to opt-in to block gap support, the idea here is that if you're using the main CSS variable, then you should expect it to behave like the CSS `gap` attribute. However, if you're using the separate row/column values, then you can be free to use the value in other properties as you like?

**Comment by ramonjd** [comment]
> In terms of the API we're effectively creating via the CSS variables, I suppose for other blocks that might want to opt-in to block gap support, the idea here is that if you're using the main CSS variable, then you should expect it to behave like the CSS gap attribute. However, if you're using the separate row/column values, then you can be free to use the value in other properties as you like?

Makes sense. I was genuinely asking in case I'd missed something. 😆 

Not a difficult change to make, and I think it would also be good to maintain the current CSS gap behaviour as well.

Thanks!

**Comment by ramonjd** [comment]
> Something I noticed when running TwentyTwentyTwo theme is that it sets in its global styles the spacing of 1.5rem in its theme.json which gets rendered out on the server to --wp--style--block-gap as a single CSS variable. I think this means that we might need additional logic in WP_Theme_JSON_Gutenberg for rendering out the body spacing variables (I think this is where it's set), to also generate the other two variables when a string value is set?

If I understand correctly, we need to automatically set the new CSS vars `--wp--style--block-[row|column]-gap`  where a theme defines  `--wp--style--block-gap ` so that any underlying CSS that relies on the individual variables (`--wp--style--block-[row|column]-gap` ) will pick up the value.

So if a particular theme sets :

```json
		"spacing": {
			"blockGap": "2em" // or "blockGap": "1.5rem 2em"
		},
```

We'd need to also set the following:

```css
body {
      ...
	--wp--style--block-gap: 2em; // or 1.5rem 2em if --wp--style--block-gap sets both
        --wp--style--block-row-gap: 2em; // or 1.5rem if --wp--style--block-gap sets both
        --wp--style--block-column-gap: 2rem;
}
```

I'm just trying to work out where the best place in include this logic and how. 

Looking at WP_Theme_JSON_Gutenberg and where the styles are generated from the properties, there don't seem to any other special cases and I don't want to start a precedent for custom logic in that particular class.

Maybe a series of parser methods external to that class. Not sure yet.

🤞 @oandregal will have some advice. :)

I'm just making notes for now. 

I'll think about this and come back with something a little more nuanced. 

Thank you!

**Comment by ramonjd** [comment]
> Looking at WP_Theme_JSON_Gutenberg and where the styles are generated from the properties, there don't seem to any other special cases and I don't want to start a precedent for custom logic in that particular class.
> 
> Maybe a series of parser methods external to that class. Not sure yet.

After a very cursory glance, maybe we could implement something along the lines of [PRESETS_METADATA](https://github.com/WordPress/gutenberg/blob/acb9ec5666db76c6b22bdfbd27e74e218f2d1030/lib/compat/wordpress-5.9/class-wp-theme-json-gutenberg.php#L96), with a collection of bootstrapping instructions for `styles[$property]` for example.

Just an example, I have no idea which properties would be appropriate. Still learning and looking.

```php
const STYLES_METADATA = array(
    array(
      'path' => array( 'spacing', 'blockGap' ),
      'css_vars' => array( '--wp--style--block-row-gap', '--wp--style--block-column-gap' ),
      'value_func' => 'gutenberg_render_blockgap_style',
   ),
);
```

**Comment by ramonjd** [comment]
> Looking at WP_Theme_JSON_Gutenberg and where the styles are generated from the properties, there don't seem to any other special cases and I don't want to start a precedent for custom logic in that particular class.
> 
> Maybe a series of parser methods external to that class. Not sure yet.
> 

I'm trying some things over at https://github.com/WordPress/wordpress-develop/pull/2038

Still early days and lots of questions. When I gather my notes and thoughts, I'll probably write up an issue to get feedback on them. 🙇

**Comment by oandregal** [comment]
Hey, haven't been able to dig deep into this PR and https://github.com/WordPress/wordpress-develop/pull/2038 My main question is whether this is targeted for WordPress 5.9 or 6.0?

If it's 6.0 we need to do a few things: 1) do not port it to wordpress-develop just yet, and 2) we can't modify code that lives in `lib/compat/wordpress-5.9`, instead we need to figure out what code needs to live in `lib/compat/wordpress-6.0` (I can help figuring out this before this ships). Related conversation at https://github.com/WordPress/gutenberg/pull/36909#issuecomment-990913068

If it's 5.9, I can spend some time digging into this and help fitting it into the current codebase.

**Comment by ramonjd** [comment]
Thanks for your thoughts @oandregal !


> haven't been able to dig deep into this PR and WordPress/wordpress-develop#2038 My main question is whether this is targeted for WordPress 5.9 or 6.0?

That's a good question. 

There is at least one bug that might benefit from a solution around axial gap support. 

- https://github.com/WordPress/gutenberg/issues/36521

I'm not sure about the severity, and therefore the need to target 5.9 or not. It appears to be something that might raise eyebrows if we do not fix it.

In that issue, @youknowriad also [proposed a way](https://github.com/WordPress/gutenberg/issues/36521#issuecomment-981459309) to resolve a conflict between theme.json and block-level block spacing settings. 

I'd imagine the code for that idea would similarly need to find a home.

> If it's 6.0 we need to do a few things: 1) do not port it to wordpress-develop just yet, and 2) we can't modify code that lives in lib/compat/wordpress-5.9, instead we need to figure out what code needs to live in lib/compat/wordpress-6.0 (I can help figuring out this before this ships). 
> If it's 5.9, I can spend some time digging into this and help fitting it into the current codebase.

Thanks very much! I'd love your advice on how to do this.

For this PR specifically, assuming the concept behind it sound, I was thinking it would helpful to be able to filter style properties after we add them to the `$declarations` collection. Maybe in [compute_style_properties](https://github.com/WordPress/wordpress-develop/blob/1787bcefbba8ba40aae9eb5999e0f5b93e0fda42/src/wp-includes/class-wp-theme-json.php#L1267) 🤷

**Comment by youknowriad** [comment]
The solution to this bug #36521 means the removal of the CSS variables (at least for everything but the root level of theme.json) which would definitely impact this PR. We should probably focus on that related issue first before introducing new CSS variables which would suffer from the same issues.

**Comment by ramonjd** [comment]
Closing this as we're not using CSS vars for block gap styles. See https://github.com/WordPress/gutenberg/pull/37360

**Pull request closed without merging**